models/tokens/solana/tokens_solana_token22_spl_transfers.sql

```diff
@@ -0,0 +1,145 @@
+ {{
+  config(
+        schema = 'tokens_solana',
+        alias = 'token22_spl_transfers',
+        materialized = 'incremental',
+        file_format = 'delta',
+        incremental_strategy = 'merge',
+        partition_by = ['block_date'],
+        incremental_predicates = [incremental_predicate('DBT_INTERNAL_DEST.block_time')],
+        unique_key = ['tx_id','outer_instruction_index','inner_instruction_index', 'block_slot']
+  )
+}}
+
+
+WITH
+base as (
+      --token2022. Most mint and account extensions still use the parent transferChecked instruction, hooks are excecuted after and interest-bearing is precalculated.
+      SELECT
+            account_source, account_destination, amount
+            , call_tx_id, call_block_time, call_block_slot, call_outer_executing_account, call_tx_signer
+            , action
+            , call_outer_instruction_index, call_inner_instruction_index
+            , fee
+            , token_version
+      FROM (
+            SELECT
+                  account_source, account_destination
+                  , bytearray_to_uint256(bytearray_reverse(bytearray_substring(call_data,1+1,8))) as amount --note that interestbearing mints have a different amount methodology, to add later
+                  , call_tx_id, call_block_time, call_block_slot, call_outer_executing_account, call_tx_signer
+                  , 'transfer' as action
+                  , call_outer_instruction_index, call_inner_instruction_index
+                  , least(
+                        cast(bytearray_to_uint256(bytearray_reverse(bytearray_substring(call_data,1+1,8))) as double)
+                              *cast(f.fee_basis as double)/10000
+                        ,f.fee_maximum) as fee --we want to take the percent fee on total amount, but not exceed the maximum fee
+                  , 'token2022' as token_version
+                  , f.fee_time
+                  , row_number() over (partition by tr.call_tx_id,  tr.call_outer_instruction_index,  tr.call_inner_instruction_index order by f.fee_time desc) as latest_fee
+            FROM {{ source('spl_token_2022_solana','spl_token_2022_call_transferChecked') }} tr
+            LEFT JOIN {{ ref('tokens_solana_fees_history') }} f ON tr.account_tokenMint = f.account_mint AND tr.call_block_time >= f.fee_time
```

Comment:
We only care about the latest fee for a transfer (per predicate `WHERE latest_fee = 1`).
Could we join to a more narrow window of fees? Depending on how ofter there are fee updates, it could be for example one day:
`... AND tr.call_block_time >= f.fee_time AND tr.call_block_time - interval '1 day' < f.fee_time`
That would limit the cost of TopNRanking operator downstream.

Reply:
Unfortunately not, because the latest fee might be set even a year or two ago.